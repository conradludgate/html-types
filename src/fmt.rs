use std::fmt::Write;

use crate::{Element, ElementList};

pub struct Formatter<'a> {
    depth: usize,
    pub(crate) buf: &'a mut (dyn Write + 'a),
}

pub type Result<T = (), E = std::fmt::Error> = std::result::Result<T, E>;

impl<'a> Formatter<'a> {
    pub(crate) fn render_tabs(&mut self) -> Result {
        for _ in 0..self.depth {
            self.buf.write_str("  ")?;
        }
        Ok(())
    }
    pub(crate) fn write_str(&mut self, s: &str) -> Result {
        self.render_tabs()?;
        self.buf.write_str(s)?;
        self.buf.write_char('\n')
    }
    pub fn open_and_close(&mut self, element: &'static str) -> Result {
        self.render_tabs()?;
        self.buf.write_char('<')?;
        self.buf.write_str(element)?;
        self.buf.write_str("/>\n")
    }
    pub fn open(&mut self, element: &'static str) -> Result<ElementFmt<'_, 'a>> {
        self.render_tabs()?;
        self.buf.write_char('<')?;
        self.buf.write_str(element)?;
        self.buf.write_str(">\n")?;
        self.depth += 1;
        Ok(ElementFmt {
            fmt: self,
            element,
        })
    }
}

#[must_use]
pub struct ElementFmt<'a, 'b> {
    pub(crate) fmt: &'a mut Formatter<'b>,
    element: &'static str,
}

impl ElementFmt<'_, '_> {
    #[inline]
    pub fn render_child(self, child: &impl Element) -> Result<Self> {
        child.render(self.fmt)?;
        Ok(self)
    }

    #[inline]
    pub fn render_children(self, children: &impl ElementList) -> Result<Self> {
        children.render(self)
    }

    pub fn close(self) -> Result {
        self.fmt.depth -= 1;
        self.fmt.render_tabs()?;
        self.fmt.buf.write_str("</")?;
        self.fmt.buf.write_str(self.element)?;
        self.fmt.buf.write_str(">\n")
    }
}

pub trait ElementExt: Element {
    fn render_to_string(&self) -> Result<String> {
        let mut buf = String::new();
        self.render(&mut Formatter {
            depth: 0,
            buf: &mut buf,
        })?;
        Ok(buf)
    }
}
impl<E: Element> ElementExt for E {}
