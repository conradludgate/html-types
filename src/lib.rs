pub use element::Element;
pub use element_list::ElementList;
use fmt::{ElementFmt, Formatter, Result};

mod element;
mod element_list;
pub mod fmt;

pub struct Body;
pub struct Div;

pub trait Tag {
    const NAME: &'static str;

    fn fmt<'a, 'b>(&self, fmt: &'a mut Formatter<'b>) -> Result<ElementFmt<'a, 'b>> {
        fmt.start(Self::NAME)
    }
}
pub trait Parent: Tag {}

impl Tag for Div {
    const NAME: &'static str = "div";
}
impl Parent for Div {}
impl Tag for Body {
    const NAME: &'static str = "body";
}
impl Parent for Body {}

pub struct WithChildren<T, C> {
    tag: T,
    children: C,
}

pub trait ChildrenExt: Parent + Sized {
    fn children<C: ElementList>(self, children: C) -> WithChildren<Self, C> {
        WithChildren {
            tag: self,
            children,
        }
    }
}
impl<P: Parent> ChildrenExt for P {}

pub struct Closed<T>(T);
pub trait TagExt: Tag + Sized {
    fn end(self) -> Closed<Self> {
        Closed(self)
    }
}
impl<T: Tag> TagExt for T {}

impl<T: Tag> Element for Closed<T> {
    fn render(&self, fmt: &mut Formatter) -> Result {
        self.0.fmt(fmt).map(|_| ())
    }
}

impl<P: Parent, C: ElementList> Element for WithChildren<P, C> {
    fn render(&self, fmt: &mut Formatter) -> Result {
        self.children.render(self.tag.fmt(fmt)?).map(|_| ())
    }
}

impl Element for str {
    fn render(&self, fmt: &mut Formatter) -> Result {
        fmt.write_str(self)
    }
}

impl Element for String {
    fn render(&self, fmt: &mut Formatter) -> Result {
        fmt.write_str(self)
    }
}

#[cfg(test)]
mod tests {
    use crate::fmt::ElementExt;

    use super::*;
    use insta::assert_snapshot;

    fn demo_template(name: &str) -> impl Element + '_ {
        Body.children((
            Div.children(["title", name]),
            Div.children(["foo", "bar"]),
            Div.end(),
            "baz",
        ))
    }
    
    #[test]
    fn demo() {
        assert_snapshot!(demo_template("hello").render_to_string().unwrap());
    }
}

