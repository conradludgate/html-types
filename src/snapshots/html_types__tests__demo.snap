---
source: src/lib.rs
expression: "demo_template(\"hello\").render_to_string().unwrap()"
---
<body>
  <div>
    title
    hello
  </div>
  <div>
    foo
    bar
  </div>
  <div/>
  baz
</body>

